safe_control_gym/controllers/mpc/mpc_acados.py

```diff
@@ -0,0 +1,391 @@
+            terminate_run_on_done (bool): Terminate the run when the environment returns done or not.
+            constraint_tol (float): Tolerance to add the the constraint as sometimes solvers are not exact.
+            output_dir (str): output directory to write logs and results.
+            additional_constraints (list): List of additional constraints
+            use_gpu (bool): False (use cpu) True (use cuda).
+            seed (int): random seed.
+            use_RTI (bool): Real-time iteration for acados.
+            use_lqr_gain_and_terminal_cost (bool): Use LQR ancillary gain and terminal cost for the MPC.
+        '''
+        for k, v in locals().items():
+            if k != 'self' and k != 'kwargs' and '__' not in k:
+                self.__dict__.update({k: v})
+        super().__init__(
+            env_func,
+            horizon=horizon,
+            q_mpc=q_mpc,
+            r_mpc=r_mpc,
+            warmstart=warmstart,
+            soft_constraints=soft_constraints,
+            soft_penalty=soft_penalty,
+            terminate_run_on_done=terminate_run_on_done,
+            constraint_tol=constraint_tol,
+            output_dir=output_dir,
+            additional_constraints=additional_constraints,
+            # compute_initial_guess_method='ipopt',  # use ipopt initial guess by default
+            use_lqr_gain_and_terminal_cost=use_lqr_gain_and_terminal_cost,
+            use_gpu=use_gpu,
+            seed=seed,
+            **kwargs
+        )
+
+        self.x_guess = None
+        self.u_guess = None
+        # acados settings
+        self.use_RTI = use_RTI
+
+    @timing
+    def reset(self):
+        '''Prepares for training or evaluation.'''
+        print(colored('Resetting MPC', 'green'))
+        super().reset()
+        self.acados_model = None
+        self.ocp = None
+        self.acados_ocp_solver = None
+        # Dynamics model.
+        self.setup_acados_model()
+        # Acados optimizer.
+        self.setup_acados_optimizer()
+        self.acados_ocp_solver = AcadosOcpSolver(self.ocp)
+
+    def setup_acados_model(self) -> AcadosModel:
+        '''Sets up symbolic model for acados.
+
+        Returns:
+            acados_model (AcadosModel): acados model object.
+        '''
+
+        acados_model = AcadosModel()
+        acados_model.x = self.model.x_sym
+        acados_model.u = self.model.u_sym
+        acados_model.name = self.env.NAME
+
+        # set up rk4 (acados need symbolic expression of dynamics, not function)
+        k1 = self.model.fc_func(acados_model.x, acados_model.u)
+        k2 = self.model.fc_func(acados_model.x + self.dt / 2 * k1, acados_model.u)
+        k3 = self.model.fc_func(acados_model.x + self.dt / 2 * k2, acados_model.u)
+        k4 = self.model.fc_func(acados_model.x + self.dt * k3, acados_model.u)
+        f_disc = acados_model.x + self.dt / 6 * (k1 + 2 * k2 + 2 * k3 + k4)
+
+        acados_model.disc_dyn_expr = f_disc
+        '''
```

Comment:
I would put this in the function docstring maybe. I found it a bit sudden and out of place here

Reply:
This has been moved to docstring.